### ISSUE TITLE
Closing browser tab should warn user about background activity

### ISSUE THREAD

**Issue by balamurugana** [minio/minio#4748]
Based on https://github.com/minio/minio/issues/4670 and https://github.com/minio/minio/issues/4494, Closing browser tab should warn user about background activity like GMail.  Currently it causes unexpected EOF on client connections at server side .

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Your Environment

* Version used (`minio version`):
* Environment name and version (e.g. nginx 1.9.1):
* Server type and version:
* Operating System and version (`uname -a`):
* Link to your project:

**Comment by deekoder** [comment]
@balamurugana  i assume this is for the error message printed on the server. However I am not sure how you want to tell the user about the background processes running when they close the browser. Couldnt we simple stop the background activity if they decided to close the tab. I think we can take the tab closure as an intent to close background processes?

**Comment by balamurugana** [comment]
@deekoder I think we would need to do exactly what GMail handles this case.

**Comment by deekoder** [comment]
Can you elaborate how gmail handles it. Is there a warning pop up on the ux for eg?

**Comment by balamurugana** [comment]
Yes. GMail shows warning popup about background tasks on tab closing.

**Comment by krisis** [comment]
@balamurugana what is the background task in the case of minio browser?

**Comment by balamurugana** [comment]
@krisis I don't know.  If you see issues https://github.com/minio/minio/issues/4670 and https://github.com/minio/minio/issues/4494, closing browser tab causes EOF client connections (continuously).  This is clear evidence that browser is doing some tasks.

**Comment by deekoder** [comment]
BTW that happens for me even without closing the browser.

**Comment by brendanashworth** [comment]
I'm looking into the issue. This only happens when using chrome.

**Comment by brendanashworth** [comment]
There are two issues here. The first is closing the browser when backgrounds tasks are occurring. This is separate from the "Error in reading from new connection" error in https://github.com/minio/minio/pull/4762. The following explanation is for the "reading from new connection" bug.

I believe the "issue" here is that Chrome is opening up connections pre-emptively, so as to make connecting to the web page faster. These EOFs need to be handled gracefully, as when a pre-emptive connection isn't used, Chrome will just give it EOF.

According to [this webpage](https://www.igvita.com/posa/high-performance-networking-in-google-chrome/#tcp-pre-connect), Google Chrome will open up TCP connections to perform the handshake before the webpage is even loaded, or the user has even told it what to do. If you type in "localh" into your Chrome browser and have visited "localhost:9000" before, Chrome will actually open up a new TCP connection to that host before you finish typing. This is to be faster. But when it actually makes the requests to the browser, it might find out that it only uses 3 out of the 5 connections it opened, or 15 out of the 20, etc. Chrome will then send an EOF on those extra connections before any data is sent, which is why `.Peek()` returns an EOF before any data is sent.

This optimization is also why you cannot reproduce the bug by reloading the page, only by typing the page in. I believe when you keep the browser continuously open, it does something similar. We should be gracefully handling these errors.

```
ERRO[0080] Error in reading from new connection [::1]:51748 at server [::1]:9000  cause=EOF source=[listener.go:127:github.com/minio/minio/pkg/http.(*httpListener).start.func2()]
```

And for reference, here's the socket Chrome opened up. Note that the outgoing port (51748) is the same.

![screen shot 2017-08-03 at 4 07 55 pm](https://user-images.githubusercontent.com/3038254/28947820-c12da4ea-7866-11e7-89f6-3ad5eac90d3e.png)

![screen shot 2017-08-03 at 4 21 21 pm](https://user-images.githubusercontent.com/3038254/28948011-dbb18fa6-7867-11e7-9324-180d4d93f0a3.png)


**tl;dr:** we should ignore these errors if they're an `io.EOF`.

**Comment by harshavardhana** [comment]
This should be fixed after #4762 merge? @brendanashworth

**Comment by brendanashworth** [comment]
Ah @harshavardhana yes that should fix the error messages. I was wrongly under the impression that this issue also dealt with warning the user if they close a browser tab while uploading a file (I think that's what was said in a meeting). This can be closed, I'll put the user warning in https://github.com/minio/minio/issues/4772.

**Comment by harshavardhana** [comment]
> Ah @harshavardhana yes that should fix the error messages. I was wrongly under the impression that this issue also dealt with warning the user if they close a browser tab while uploading a file (I think that's what was said in a meeting). This can be closed, I'll put the user warning in #4772.

Issue moved to #4772 - closing this bug to be handled in newux.
